### PR TITLE
[processing] algorithm for fixing geometries with new makeValid() call

### PR DESCRIPTION
**Pull request by alexbruy**

### CODE REVIEW

**Comment by wonder-sk** [comment]
Hi @alexbruy really great to see a processing alg using makeValid()  - woohoo!

Is the algorithm going work in case of changes to the geometry type? Namely:
- invalid polygon turned into a multipolygon
- invalid polygon turned into a geometry collection (e.g. polygon + linestring)
- invalid polygon collapsing to a lower dimension (linestring or point)

As far as I understand, output to GML is very flexible, but other output formats may fail to write the results? I was thinking if we should introduce two extra options, something like:
- ignore geometry parts of different dimension
- explode multi-part geometries into multiple features

**Comment by alexbruy** [comment]
@wonder-sk right now it is very basic and does not handle cases you mentioned. Will work on your suggestions

**Comment by nyalldawson** [comment]
> explode multi-part geometries into multiple features

Wouldnt it be more predictable to just always output multi geoms? And then people can build into their models a spilt to single parts alg if desired.

**Comment by nirvn** [comment]
+1 for predictable always-multi output; I've implemented that behavior in several vector overlay algorithms a few months ago.

**Comment by pcav** [comment]
Multi are ok for polygons and lines. Multipoints are often a source of problems.

**Comment by nyalldawson** [comment]
Is there any valid use case for running makevalid on a point layer?

**Comment by wonder-sk** [comment]
Points and multi-points are always valid. (not sure if they are considered as valid with inf / NaN coordinates, but those we can't fix anyway)

I think let's merge the PR as is and then sort out any extra options if needed - they were meant as workarounds for some common formats (e.g. shp will not like geometry collection for polygon layer type, and postgis will not like multi-polygon for polygon layer type) but these are maybe better addressed in a more generic way...?

**Comment by alexbruy** [comment]
Algorithm now outputs multi-geometry layer. Also I tried to add handling of GeometryCollections

File: python/plugins/processing/algs/qgis/FixGeometry.py [hunk 1]
```diff
@@ -68,6 +71,19 @@ def processAlgorithm(self, feedback):
...
                     ProcessingLog.addToLog(ProcessingLog.LOG_WARNING,
                                            'makeValid failed for feature {}'.format(inputFeature.id()))
+
+                if outputGeometry.wkbType() == QgsWkbTypes.Unknown or QgsWkbTypes.flatType(outputGeometry.geometry().wkbType()) == QgsWkbTypes.GeometryCollection:
+                    tmpGeometries = outputGeometry.asGeometryCollection()
+                    for g in tmpGeometries:
+                        if g.type() == inputFeature.geometry().type():
+                            try:
+                                outputFeature.setGeometry(QgsGeometry(g))
+                                writer.addFeature(outputFeature)
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/FixGeometry.py]: Does writer.addFeature automatically convert to multipart when required?

Comment (alexbruy) [inline python/plugins/processing/algs/qgis/FixGeometry.py]: Should be fixed now

**Comment by alexbruy** [comment]
Any objections to merge?

**Comment by pcav** [comment]
Not from me, on the contrary!

**Comment by nyalldawson** [comment]
I think it just needs a help text added with a special note saying it will drop m values. But otherwise +1 from me.

File: python/plugins/processing/algs/qgis/FixGeometry.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+class FixGeometry(GeoAlgorithm):
+
+    INPUT = 'INPUT'
+    OUTPUT = 'OUTPUT'
+
+    def defineCharacteristics(self):
+        self.name, self.i18n_name = self.trAlgorithm('Fix geometries')
+        self.group, self.i18n_group = self.trAlgorithm('Vector geometry tools')
+
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/FixGeometry.py]: I'd also suggest adding tags here for 'repair','invalid','geometry'

**Comment by alexbruy** [comment]
Manually merged. Tags and short help also added.

**Pull request closed without merging**